cobalt-ast/src/cfg.rs

```diff
@@ -704,10 +756,104 @@ impl<'a, 'ctx> Cfg<'a, 'ctx> {
                 }
             }
         }
+
         errs.sort_unstable();
         errs.dedup_by_key(|m| m.loc);
+
+        let mut to_return = errs
+            .iter()
+            .map(
+                |DoubleMove {
+                     name,
+                     loc,
+                     prev,
+                     guaranteed,
+                 }| CobaltError::DoubleMove {
+                    loc: *loc,
+                    prev: *prev,
+                    name: name.clone(),
+                    guaranteed: *guaranteed,
+                },
+            )
+            .collect::<Vec<_>>();
+
+        to_return.append(&mut self.validate_linear_types(ctx));
+        to_return
+    }
+
+    fn validate_linear_types(&self, ctx: &'a CompCtx<'ctx>) -> Vec<cobalt_errors::CobaltError> {
+        let mut errs = Vec::new();
+
+        // Go through each variable, see if it's a linear type.
+        //
+        // - 1: We check if it has been moved at least once by verifying that the variable
+        // appears in the moves list.
+        unsafe {
+            ctx.with_vars(|v| {
+                v.symbols.iter().for_each(|(n, v)| {
+                    let is_linear_type = {
+                        let to_return: bool;
+
+                        let nominal_behavior = |type_name: &String| {
+                            if let Some((_, _, _, nom_info)) = ctx.nominals.borrow().get(type_name)
+                            {
+                                nom_info.is_linear_type
+                            } else {
+                                false
+                            }
+                        };
+
+                        // See if it's a nominal type.
+                        match &v.0.data_type {
+                            Type::Nominal(type_name) => {
+                                to_return = nominal_behavior(type_name);
+                            }
+
+                            Type::Mut(ref boxed_type) => {
+                                if let Type::Nominal(ref type_name) = **boxed_type {
+                                    to_return = nominal_behavior(type_name);
+                                } else {
+                                    to_return = false;
+                                }
+                            }
+
+                            _ => {
+                                to_return = false;
+                            }
+                        }
+
+                        to_return
+                    };
+
+                    if is_linear_type {
+                        // 1
+                        let vars = self
+                            .blocks
+                            .iter()
+                            .flat_map(|v| &v.moves)
+                            .map(|m| for_both!(m, m => &(**m).name));
+
+                        let mut moved = false;
+                        for var in vars {
```

Comment:
This only checks if a variable is moved on *one* path, not all of them. An unused variable warning could be implemented, but I wouldn't put it on `@must_use`